File: backend/app/auth/roles.py
"""Role hierarchy and helpers.

A grant of a higher role implicitly satisfies any lower-role permission
check. `effective_roles(user)` returns the full set of roles a user
satisfies based on their explicit grants.

If this hierarchy changes, mirror the change in `frontend/src/lib/roles.ts`.
"""
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from backend.app.db.models import User

ADMIN = "admin"
DEPARTMENT_MANAGER = "department_manager"
PROJECT_EDITOR = "project_editor"
VIEWER = "viewer"

# Each key maps to the set of roles that key implicitly satisfies.
ROLE_HIERARCHY: dict[str, frozenset[str]] = {
    ADMIN: frozenset({ADMIN, DEPARTMENT_MANAGER, PROJECT_EDITOR, VIEWER}),
    DEPARTMENT_MANAGER: frozenset({DEPARTMENT_MANAGER, PROJECT_EDITOR, VIEWER}),
    PROJECT_EDITOR: frozenset({PROJECT_EDITOR, VIEWER}),
    VIEWER: frozenset({VIEWER}),
}

KNOWN_ROLES: frozenset[str] = frozenset(ROLE_HIERARCHY.keys())


def effective_roles_for(granted: list[str] | set[str] | frozenset[str]) -> frozenset[str]:
    """Expand explicit role grants into the full effective role set."""
    result: set[str] = set()
    for role in granted:
        result.update(ROLE_HIERARCHY.get(role, frozenset()))
    return frozenset(result)


def effective_roles(user: "User") -> frozenset[str]:
    """Return the effective role set for a user based on their user_roles rows."""
    return effective_roles_for([ur.role_id for ur in user.user_roles])

File: frontend/src/lib/roles.ts
import type { BadgeTone } from "@/components/Badge"

// Mirror of backend `backend/app/auth/roles.py`. Keep these two in sync.
// Permission ladder (highest → lowest):
//   admin > department_manager > project_editor > viewer
// A grant of a higher role implicitly satisfies any lower-role check.

export const ROLE_HIERARCHY: Record<string, ReadonlyArray<string>> = {
  admin: ["admin", "department_manager", "project_editor", "viewer"],
  department_manager: ["department_manager", "project_editor", "viewer"],
  project_editor: ["project_editor", "viewer"],
  viewer: ["viewer"],
}

export function effectiveRoles(roles: ReadonlyArray<string>): Set<string> {
  const out = new Set<string>()
  for (const r of roles) {
    for (const e of ROLE_HIERARCHY[r] ?? []) {
      out.add(e)
    }
  }
  return out
}

export function hasRole(
  granted: ReadonlyArray<string>,
  required: string,
): boolean {
  return effectiveRoles(granted).has(required)
}

// True when the user holds `role` (or a hierarchy-superior role) via a
// NULL-dept grant. Mirrors backend `is_org_role(user, role)`. Today's
// allowed org-scope grants: (admin, NULL) and (viewer, NULL).
export function isOrgRole(
  grants: ReadonlyArray<{ role_id: string; department_id: string | null }>,
  role: string,
): boolean {
  return grants.some(
    (g) =>
      g.department_id === null &&
      (ROLE_HIERARCHY[g.role_id] ?? []).includes(role),
  )
}

const ROLE_LABELS: Record<string, string> = {
  admin: "Admin",
  department_manager: "Department Manager",
  project_editor: "Project Editor",
  viewer: "Viewer",
}

export function roleLabel(roleId: string): string {
  return ROLE_LABELS[roleId] ?? roleId
}

const ROLE_TONES: Record<string, BadgeTone> = {
  admin: "indigo",
  department_manager: "blue",
  project_editor: "emerald",
  viewer: "slate",
}

export function roleTone(roleId: string): BadgeTone {
  return ROLE_TONES[roleId] ?? "slate"
}
